### PR TITLE
Activate k8s on no logs_config.container_collect_all

### PR DESCRIPTION
**Pull request by achntrl**
### What does this PR do?

This PR allow k8s logs integration even when `logs_config.container_collect_all` is not enabled. In that case, only pods with logs annotation are tailed

### Motivation

Coherence with the docker log integration behavior

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: pkg/logs/input/container/launcher.go [hunk 1]
```diff
@@ -24,26 +24,18 @@ import (
...
-		if err == nil {
-			return launcher
-		}
-		log.Warnf("Could not setup the docker launcher: %v", err)
+	// attempt to initialize a docker launcher
+	launcher, err := docker.NewLauncher(sources, services, pipelineProvider, registry)
+	if err == nil {
+		return launcher
 	}
+	log.Warnf("Could not setup the docker launcher: %v", err)
```
Comment (ajacquemot) [inline pkg/logs/input/container/launcher.go]: Wondering whether this logs should be an info

File: pkg/logs/input/container/launcher.go [hunk 2]
```diff
@@ -24,26 +24,18 @@ import (
 // the launcher will attempt to initialize a kubernetes launcher which will detect and tail all the logs files localized
 // in '/var/log/pods' of all the containers running on the kubernetes cluster.
 func NewLauncher(collectAll bool, sources *config.LogSources, services *service.Services, pipelineProvider pipeline.Provider, registry auditor.Registry) restart.Restartable {
-	switch {
```
Comment (ajacquemot) [inline pkg/logs/input/container/launcher.go]: Do we want to add a log that explains the strategy here ? i.e. try docker, fallback to kubernetes, finally noop ?

File: pkg/logs/input/kubernetes/launcher.go [hunk 3]
```diff
@@ -61,6 +70,14 @@ func NewLauncher(sources *config.LogSources, services *service.Services) (*Launc
 	return launcher, nil
 }
 
+func isAvailable() bool {
+	if _, err := os.Stat(podsDirectoryPath); os.IsNotExist(err) {
```
Comment (ajacquemot) [inline pkg/logs/input/kubernetes/launcher.go]: what if the error is a access denied ?

File: pkg/logs/input/container/launcher.go [hunk 4]
```diff
@@ -24,26 +24,18 @@ import (
 // the launcher will attempt to initialize a kubernetes launcher which will detect and tail all the logs files localized
```
Comment (ajacquemot) [inline pkg/logs/input/container/launcher.go]: this needs to be updated ` // When a docker launcher can not be initialized properly and when the log collection is enabled for all containers,`

Comment (ajacquemot) [reply pkg/logs/input/container/launcher.go]: as well as the header of this method

File: pkg/logs/input/kubernetes/launcher.go [hunk 5]
```diff
@@ -61,6 +70,14 @@ func NewLauncher(sources *config.LogSources, services *service.Services) (*Launc
 	return launcher, nil
 }
 
+func isAvailable() bool {
```
Comment (ajacquemot) [inline pkg/logs/input/kubernetes/launcher.go]: nits: `isIntegrationAvailable` ? or add a comment please

File: pkg/logs/input/kubernetes/launcher.go [hunk 6]
```diff
@@ -22,6 +23,8 @@ import (
 // The path to the pods log directory.
 const podsDirectoryPath = "/var/log/pods"
 
+var collectAllDisabledError = fmt.Errorf("%s disabled", config.ContainerCollectAll)
```
Comment (ajacquemot) [inline pkg/logs/input/kubernetes/launcher.go]: as this error seems to be duplicated, we should probably have only one public ?

File: pkg/logs/input/kubernetes/launcher.go (hunk 3 above)
Comment (achntrl) [reply pkg/logs/input/kubernetes/launcher.go]: Agreed, we probably don't want to launch on any Stat error

File: pkg/logs/input/container/launcher.go [hunk 7]
```diff
@@ -17,33 +17,30 @@ import (
...
 
-// NewLauncher returns a new container launcher,
-// by default returns a docker launcher that uses the docker socket to collect logs.
+// NewLauncher returns a new container launcher.
+// By default, it returns a docker launcher that uses the docker socket to collect logs.
 // The docker launcher can be used both on a non kubernetes and kubernetes environment.
-// When a docker launcher can not be initialized properly and when the log collection is enabled for all containers,
-// the launcher will attempt to initialize a kubernetes launcher which will detect and tail all the logs files localized
-// in '/var/log/pods' of all the containers running on the kubernetes cluster.
+// When a docker launcher cannot be initialized properly. The launcher will attempt to
```
Comment (ajacquemot) [inline pkg/logs/input/container/launcher.go]: nits: `properly. The launcher` -> `properly, the launcher`

File: pkg/logs/input/container/launcher.go [hunk 8]
```diff
@@ -17,33 +17,30 @@ import (
...
-		log.Warnf("Could not setup the kubernetes launcher: %v", err)
-	default:
-		launcher, err := docker.NewLauncher(sources, services, pipelineProvider, registry)
-		if err == nil {
-			return launcher
-		}
-		log.Warnf("Could not setup the docker launcher: %v", err)
+	// attempt to initialize a docker launcher
+	log.Info("Trying to initialize docker launcher")
+	launcher, err := docker.NewLauncher(sources, services, pipelineProvider, registry)
```
Comment (ajacquemot) [inline pkg/logs/input/container/launcher.go]: Is there some kind of errors that could be worth logging as warning or error ?

File: pkg/logs/input/container/launcher.go [hunk 9]
```diff
@@ -17,33 +17,30 @@ import (
...
+	if err == nil {
+		log.Info("Docker launcher initialized")
+		return launcher
 	}
+	log.Infof("Could not setup the docker launcher: %v", err)
+
+	// attempt to initialize a kubernetes launcher
+	log.Info("Trying to initialize kubernetes launcher")
+	kubernetesLauncher, err := kubernetes.NewLauncher(sources, services, collectAll)
```
Comment (ajacquemot) [inline pkg/logs/input/container/launcher.go]: Same here?

**Review by ajacquemot — APPROVED** [review]
Left just two comments, feel free to discard them if they are not relevant.

**Comment by achntrl** [comment]
Let's merge it as it and keep an eye on how often it appears. I don't think we will have much errors other than "we not should activate this integration because resources aren't here" because most of the time the agent is run by the root process of the container.

**Pull request merged**